SSR adjustments

Allcharles opened the pull request:
# Adjustments to Server Side Rendering

Attempting to speed up SSR bundles, and fix some of the SSR only errors which occur in the website.

## Changes

- ~Set 1 second timeout on SSR API requests~
- Adjusted home page to handle mobile device screens
- Set default background colour for home page when assets fail to load
- Prevent use of line-trunction on server because of https://github.com/DiZhou92/ngx-line-truncation/issues/49
- Prevent use of google maps on server
- Disabled CMS on SSR because insertion requires access to `document`
- Enhanced configuration file checking
- Removed edge tests as they should produce the same results as chrome
- Forced `git-lfs` files to download to hopefully fix assets not being available in docker container
- Updated readme with documentation on server side rendering
- Removed snazzy-info-window dependency
- Added fontawsome icon styles to `angular.json` styles to fix large icons bug when initially rendered. Example of bug shown below:
![image](https://user-images.githubusercontent.com/3955116/122704883-f0655700-d297-11eb-990e-35e8c8af915b.png)
- Cache SSR files in express server
- Fixed bug where `ngx-loading-bar` was causing the server side renderer to wait 3 seconds before returning to the client
- Set max timeouts for HTTP requests for both website (10 seconds) and server (100ms)
- Cache HTTP requests on SSR

## Results

Tests run on latest version of Firefox, using the throttling settings from developer tools

### Before

#### Regular 3G

**Home Page**

![image](https://user-images.githubusercontent.com/3955116/122705060-4df9a380-d298-11eb-85fa-e1ddd8e056dc.png)

**Projects Page**

![image](https://user-images.githubusercontent.com/3955116/122705124-71245300-d298-11eb-8ea8-f5366ff8d3c2.png)

### After

#### Regular 3G

**Home Page**

![image](https://user-images.githubusercontent.com/3955116/122713551-31b23280-d2a9-11eb-8f35-3952195dcfe1.png)

**Projects Page**

![image](https://user-images.githubusercontent.com/3955116/122705550-556d7c80-d299-11eb-90b7-d8db0125b483.png)

## Issues

Addresses #1108 
Closes #1106 

## Visual Changes

### Mobile home title

![image](https://user-images.githubusercontent.com/3955116/121292648-a8f0d980-c92d-11eb-8e65-503a4ac80c33.png)

### Normal home title

![image](https://user-images.githubusercontent.com/3955116/121292681-b3ab6e80-c92d-11eb-8bc9-6aebe6982894.png)

### Default home background

![image](https://user-images.githubusercontent.com/3955116/121292579-8d85ce80-c92d-11eb-86a2-3dd7636e0f11.png)


## Final Checklist

- [ ] Assign reviewers if you have permission
- [ ] Assign labels if you have permission
- [ ] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

Allcharles committed:
Cleanup error messages during SSR

Allcharles committed:
Update README.md

Allcharles committed:
Update configation file checker

Allcharles committed:
Fix lint errors

atruskie commented on src/app/helpers/embedGoogleMaps/embedGoogleMaps.ts:
```diff
@@ -11,12 +11,16 @@ let node: HTMLScriptElement;
     googleMapsUrl += "?key=" + key;
   }
 
+  if (!document) {
```
not the isSSR function?

atruskie reviewed (approved):
The `isSSR` function seems like it could be defined... yup there we go: https://angular.io/api/common/isPlatformBrowser

Also, a question: did you try lower timeouts?  From a short profile of API requests, it seems most requests average 30 - 60 ms. Can we get our timeout lower?

It'd be good to put these timeout values in our environment.json so we can tune our servers.

If a request fails during server side rendering, will it be attempted again when the client is hydrated in the browser?

Allcharles committed:
Shorten ssr timeout duration

Allcharles committed:
Simplify angular bootstrap imports for tree shaking

Allcharles committed:
Fix 3 second delay on ssr

Rxjs `delay` operator on the loading bar for http requests caused zonejs to wait for the first trigger of the observable before returning the build. This was fixed by changing the delay to use the global config for ngx-loading. ngx-loading disables its use in SSR to prevent this specific issue.

Allcharles committed:
Add caching and default timeout

Allcharles committed:
Update error handler tests

Allcharles committed:
Remove snazzy-info-window styles

Allcharles committed:
Created IS_SERVER_PLATFORM injection token

Removed isSsr function

Fix broken module deps

Allcharles committed:
Force lfs file downloads in CI

atruskie commented on server.ts:
```diff
@@ -86,7 +92,7 @@ function run(configPath: string): void {
   const server = app(configPath);
   server.listen(port, "0.0.0.0", () => {
-    console.log(`Node Express server listening on http://0.0.0.0:${port}`);
+    console.log(`Node Express server listening on http://localhost:${port}`);
```
given we intend for this to be in a container, it should almost always be 0.0.0.0. Is there a reason it's not?

atruskie commented on server.ts:
```diff
@@ -42,11 +44,15 @@ export function app(path: string): express.Express {
-    })
+    // Cache results as they should not change often for unauthenticated users
+    withCache(
+      new LRUCache({ maxAge: 10 * 60_000, maxSize: 100 }),
```
nit: add comment about units and calculated size. I assume it is 10 minutes?

atruskie commented on src/app/app.component.ts:
```diff
@@ -26,24 +24,13 @@ export class AppComponent extends withUnsubscribe() implements OnInit {
   public ngOnInit() {
     this.title.setTitle(this.config.values.brand.name);
     this.fullscreen = true;
-
-    // Delay showing loading bar
-    this.delayedProgress$ = this.loader.value$.pipe(
-      delay(3000),
-      withLatestFrom(this.loader.value$),
-      map((v) => v[1])
-    );
```
avoiding the problem altogether, smart

atruskie commented on src/app/components/shared/cms/cms.component.ts:
```diff
@@ -41,6 +46,11 @@ export class CmsComponent extends withUnsubscribe() implements OnInit {
   public ngOnInit() {
     this.loading = true;
 
+    // Don't attempt to load CMS data if currently we are in SSR
+    if (this.isServer) {
+      return;
+    }
+
```
with ssr request cache then we may not need to not do this. It would be better if we could serve CMS content.

What the perf difference here, with this on/off/with cache?

Allcharles replied:
The issue with CMS is not loading data, but the actual way in which we insert the content into the page. A solution would require a rewrite of how it works

Allcharles replied:
I have made an issue tracking this here: #1163

atruskie commented on src/app/services/baw-api/api.interceptor.service.ts:
```diff
@@ -127,6 +128,15 @@ export class BawApiInterceptor implements HttpInterceptor {
       };
+    } else if (response.status === 0) {
+      // Timeout library sets status to 0 if timed out
+      // https://github.com/IKatsuba/ngx-ssr/issues/397
```
its non-standard but conventionally cancelled requests are returned with 0

https://stackoverflow.com/a/29907525/224512
https://www.ibm.com/support/pages/what-causes-status-code-0-indicating-reqcancelled

a very detailed answer: https://stackoverflow.com/a/26451773/224512

Allcharles committed:
Using pngcheck to validate LFS asset

Allcharles committed:
Update tests to validate LFS and corruption

Removed LFS download to check tests fail

Allcharles committed:
Re-enable LFS downloads

Allcharles committed:
Fix review comments
